12-offer-encoding.md

```diff
@@ -0,0 +1,840 @@
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnNonce"||TLV1,1)
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnNonce"||TLV1,2)
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnNonce"||TLV1,3)
+
+Assume L1 < L1nonce, L2 > L2nonce and L3 > L3nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightninginvoicesignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice_request: readers will request an invoice
+(or multiple) based on the offer.  An offer can be much longer-lived than a
+particular invoice, so it has some different characteristics; in particular the amount can be in a non-lightning currency.  It's
+also designed for compactness to fit inside a QR code easily.
+
+Note that the non-signature TLV elements get mirrored into
+invoice_request and invoice messages, so they each have specific and
+distinct TLV ranges.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`offer_issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 20 (`offer_quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 22 (`offer_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 0.
+  - MUST set `offer_node_id` to the node's public key to request the invoice from.
```

Comment:
We should make `offer_node_id` implicit in case `offer_paths` is set.
There was some discussion about it at some point and everyone seemed to agree but it wasn't added here.

```suggestion
  - MUST set `offer_node_id` or `offer_paths`.
    - if `offer_node_id` is set, it is the node's public key to request the invoice from.
    - if `offer_node_id` is not set, the last `blinded_node_id` of the first `path` is used instead.
```

Reply:
I can't recall the discussion. What would be used to sign the invoice?

Reply:
When receiving a message through a blinded path, we need to derive the blinded secret key to be able to decrypt the onion. That's the key we would use to sign the invoice.

Reply:
Actually, as @t-bast has suggested, the implied node id shouldn't be the `blinded_node_id` from the first path but the one from the path that is used to to request the invoice.

```
A writer of an offer:
- MUST set `offer_node_id` or `offer_paths`

A writer of an invoice:
- if the invoice request contains a `offer_node_id`, MUST sign the invoice with `offer_node_id`.
- if not, MUST sign the invoice with the last `blinded_node_id` of the path that was used to request the invoice.
```

Reply:
Looking through our implementation and the spec, it seems presence of `offer_node_id` in an `invoice` determines if it is for an `invoice_request for an offer` or otherwise an `invoice_request without an offer`. And this in turn determines how some of the TLVs are validated when reading an `invoice`.

Reply:
Can't you just check for `invoice.offer_node_id != null || invoice.offer_paths != null` instead of just `invoice.offer_node_id != null`? It's a one line change.

Reply:
I'm not sure I understand how that helps. Or maybe I'm misunderstanding the proposal. IIUC, we want to support reading three variations of an offer:
```
- WITH offer_node_id and WITH offer_paths
- WITH offer_node_id and WITHOUT offer_paths
- WITHOUT offer_node_id and WITH offer_paths
```
But the third variation is what is used for a refund (i.e., "invoice_request without an offer"). So when receiving an `invoice` WITHOUT `offer_node_id` and WITH `offer_paths`, we can't tell if it is for an offer or a refund.

Reply:
What I'm saying is that if `offer_paths` is set, then there must be an offer and it can't be a refund.

Reply:
Reviewing the Discord discussion, @t-bast proposed changing MUST to MAY in order to still support the first variation. But it looks like your proposal disallows it.

Reply:
> What I'm saying is that if `offer_paths` is set, then there must be an offer and it can't be a refund.

Hmmm... that's not my understanding of a refund but it may be because of https://github.com/lightning/bolts/pull/798/files#r1576805027.

Reply:
> Reviewing the Discord discussion, @t-bast proposed changing MUST to MAY in order to still support the first variation. But it looks like your proposal disallows it.

I don't understand what you think is disallowed in my proposal. Everything that was allowed is still allowed, I'm just allowing a new case that was previously forbidden.

```
                         | offer_paths is set                                                             | offer_paths is not set
-------------------------+--------------------------------------------------------------------------------+------------------------------------------------------------------
offer_node_id is set     | use offer_paths to request an invoice signed with offer_node_id                | use offer_node_id to request an invoice signed with offer_node_id
-------------------------+--------------------------------------------------------------------------------+------------------------------------------------------------------
offer_node_id is not set | use offer_paths to request an invoice signed with the last node id of the path | refund
```

Reply:
Chatted with @thomash-acinq offline. Here's where we landed:

Currently, the spec states for an `invoice_request`:
```
The writer:
 - otherwise (not responding to an offer):
   - MUST set (or not set) `offer_description`, `offer_absolute_expiry`, `offer_paths` and `offer_issuer` as it would for an offer.
```
We should remove `offer_paths` from this and use something like a new `invreq_paths` TLV instead and disallow `offer_paths` in order to differentiate an invoice for offer from one for a refund.

@rustyrussell What do you think?

Reply:
Yes, I think we need a new invreq_paths for this case.  How badly does this break everyone?